x-pack/platform/plugins/shared/onechat/server/services/agents/modes/default/graph.ts

```diff
@@ -63,8 +69,35 @@ export const createAgentGraph = ({
   structuredOutput?: boolean;
   outputSchema?: Record<string, unknown>;
   processedConversation: ProcessedConversation;
+  onechatToLangchainIdMap: ToolIdMapping;
 }) => {
-  const toolNode = new ToolNode<BaseMessage[]>(tools);
+  const initialize = async (state: StateType) => {
+    const restoredState: Partial<StateType> = {};
+    let actions: ResearchAgentAction[] = [];
+
+    const lastRound = processedConversation.previousRounds.length
+      ? processedConversation.previousRounds[processedConversation.previousRounds.length - 1]
+      : undefined;
+
+    if (lastRound?.status === ConversationRoundStatus.awaitingPrompt) {
+      actions = roundToActions({ round: lastRound, toolIdMapping: onechatToLangchainIdMap });
+      restoredState.resumeToStep = steps.executeTool;
+    }
+
+    if (lastRound?.state) {
+      restoredState.currentCycle = lastRound.state.agent.current_cycle;
+      restoredState.errorCount = lastRound.state.agent.error_count;
+    }
+
+    return { ...restoredState, mainActions: [...actions] };
+  };
+
+  const initializeEdge = async (state: StateType) => {
+    if (state.resumeToStep === steps.executeTool) {
+      return steps.executeTool;
+    }
+    return steps.researchAgent;
+  };
```

Comment:
The resume logic, which restores the graph's state to where it was when we interrupted it.

(and yes, not using langgraph's snapshot mechanism is voluntary. I want us to be able to "resume" just from the conversation object as long as we can)